### PR TITLE
FE-789 Cypress tests for Summary Report

### PR DESCRIPTION
**Pull request by kchu1337**
### What Changed
 - Added integration tests for Summary Report.
    - Changing date range
    - Adding filters
    - Changing metrics
    - Changing group by(see notes)

### How To Test
 - run `npm run test-integration` and `npm run test-integration-headless`
 - Uncomment out the 4th test and confirm that it works locally.

### Notes
 - I commented out one test because it was working locally but failing on travis

### CODE REVIEW

File: src/pages/reports/components/ReportOptions.js [hunk 1]
```diff
@@ -32,71 +31,88 @@ export class ReportOptions extends Component {
...
-          <Grid>
-            <Grid.Column xs={12} md={6}>
-              <div className={styles.FieldWrapper}>
-                <DatePicker
-                  {...reportOptions}
-                  relativeDateOptions={RELATIVE_DATE_OPTIONS}
-                  disabled={reportLoading}
-                  onChange={refreshReportOptions}
-                  roundToPrecision={true}
+      <div data-id="report-options">
```
Comment (kchu1337) [inline src/pages/reports/components/ReportOptions.js]: Just added this. Rest is prettier

File: cypress/tests/integration/analytics/SummmaryReport.spec.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
+/// <reference types="Cypress" />
```
Comment (nicklemmon) [inline cypress/tests/integration/analytics/SummmaryReport.spec.js]: If you rebase with `master` this shouldn't be necessary in your spec files anymore.

Also, just for consistency, should these be named like regular JS files? Since this isn't referring to a component. With git, case changes are kind of weird. You end up having to delete the file, commit that, then re-create it in order for any difference to be tracked.

File: cypress/tests/integration/analytics/SummmaryReport.spec.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    const getDatePickerText = startDateTime => `${startDateTime} – ${now}`; //Use this dash: '–' ¯\_(ツ)_/¯
+    cy.clock(timestamp);
+    cy.visit('/reports/summary');
+
+    const dayAgo = momentDateTime
+      .subtract(1, 'day')
+      .local()
+      .format('MMM Do YYYY h:mma');
+    cy.url().should('not.include', 'range=7days');
+    cy.get('[data-id="report-options"]').within(() => {
```
Comment (nicklemmon) [inline cypress/tests/integration/analytics/SummmaryReport.spec.js]: Weird tabbing here? Not sure why Prettier doesn't handle this!

File: cypress/tests/integration/analytics/SummmaryReport.spec.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+      cy.get('input').should('have.value', getDatePickerText(dayAgo));
+      cy.get('select').select('7days');
```
Comment (nicklemmon) [inline cypress/tests/integration/analytics/SummmaryReport.spec.js]: It would be better to use `cy.findByLabelText()` - I actually added some labels to those elements that are hidden at the moment. The labels are:

`'Broad Date Range'`
`'Narrow Date Range'`

File: cypress/tests/integration/analytics/SummmaryReport.spec.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+      requestAlias: 'getDataCampaign',
+    });
+
+  });
+
+  it('Handles changing the date range correctly', () => {
+    const timestamp = 1580392800000; //01/30/2020 @ 2:00pm (UTC)
+    const momentDateTime = Cypress.moment(timestamp);
+    const now = momentDateTime.local().format('MMM Do YYYY h:mma');
+    const getDatePickerText = startDateTime => `${startDateTime} – ${now}`; //Use this dash: '–' ¯\_(ツ)_/¯
```
Comment (nicklemmon) [inline cypress/tests/integration/analytics/SummmaryReport.spec.js]: Nice!

File: cypress/tests/integration/analytics/SummmaryReport.spec.js [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    cy.url().should('include', 'range=7days');
+    cy.get('[data-id="report-options"]').within(() => {
+      cy.get('input').should('have.value', getDatePickerText(sevenDaysAgo));
+    });
+  });
+
+  it('Handles applying filters correctly', () => {
+    cy.visit('/reports/summary');
+    cy.get('[data-id="report-options"]').within(() => {
+      cy.findByPlaceholderText('Filter by domain, campaign, etc').type('sparkpost');
```
Comment (nicklemmon) [inline cypress/tests/integration/analytics/SummmaryReport.spec.js]: This field needs a `<label>` with a connecting `id` and `htmlFor` attribute. Would you mind adding one instead of using placeholder text?

You can then style with label with:
```scss
.Label {
  @include visually-hidden()
}
```

File: cypress/tests/integration/analytics/SummmaryReport.spec.js [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+  it('Handles changing metrics correctly', () => {
+    cy.visit('/reports/summary');
+    cy.get('[data-id="summary-chart"]').within(() => {
+      cy.findByText('Accepted').should('be.visible');
+      cy.findByText('Select Metrics').click();
+    });
+
+    cy.findByText('Select up to 5 metrics').should('be.visible');
+    cy.findByLabelText('Accepted').uncheck({ force: true });
+    cy.findByLabelText('Accepted').uncheck({ force: true });
```
Comment (nicklemmon) [inline cypress/tests/integration/analytics/SummmaryReport.spec.js]: Mind adding a comment explaining why this was needed?

File: cypress/tests/integration/analytics/SummmaryReport.spec.js [hunk 8]
```diff
@@ -0,0 +1,107 @@
+/// <reference types="Cypress" />
+
+describe('Summary Report page', () => {
```
Comment (nicklemmon) [inline cypress/tests/integration/analytics/SummmaryReport.spec.js]: It would be good to add more tests for empty states and error responses from the server. There are a few examples of that in existing Cypress tests.

Comment (nicklemmon) [reply cypress/tests/integration/analytics/SummmaryReport.spec.js]: I also don't see a test for the 'Share' button?

File: cypress/tests/integration/analytics/SummmaryReport.spec.js [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+      cy.findByText('Accepted').should('not.be.visible');
+      cy.findByText('Injected').should('be.visible');
+    });
+  });
+
+  //Breaks in Travis CI
+  /*  it('Handles changing group by metric in the table', () => {
+      cy.visit('/reports/summary');
+      cy.get('[data-id="summary-table"]').within(() => cy.get('select').select('campaign'));
+      cy.wait('@getDataCampaign');
```
Comment (nicklemmon) [inline cypress/tests/integration/analytics/SummmaryReport.spec.js]: I think I figured it out! Try waiting on `@dataGetDeliverability` - looks like the wait is on the wrong request maybe?

File: cypress/tests/integration/analytics/SummmaryReport.spec.js (hunk 8 above)
Comment (kchu1337) [reply cypress/tests/integration/analytics/SummmaryReport.spec.js]: I wasn't going add a test for share since it seems fairly compartmentalized and I didn't want to add too many tests since travis runs are slow.

File: cypress/tests/integration/analytics/SummmaryReport.spec.js (hunk 3 above)
Comment (kchu1337) [reply cypress/tests/integration/analytics/SummmaryReport.spec.js]: What do you mean? The next 2 liens are inside the `within` callback

File: cypress/tests/integration/analytics/SummmaryReport.spec.js (hunk 9 above)
Comment (kchu1337) [reply cypress/tests/integration/analytics/SummmaryReport.spec.js]: Still failing, I'll just fix this when I can get the video to see the exact problem.

File: src/pages/reports/components/tests/Typeahead.test.js [hunk 10]
```diff
@@ -2,47 +2,53 @@ import React from 'react';
...
       refreshTypeaheadCache: jest.fn(() => Promise.resolve()),
       placeholder: '',
-      pattern: null
+      pattern: null,
     };
 
     downshiftHelpers = {
       getInputProps: jest.fn(),
       getItemProps: jest.fn(),
+      getLabelProps: jest.fn(),
```
Comment (kchu1337) [inline src/pages/reports/components/tests/Typeahead.test.js]: only thing I added

File: src/pages/reports/components/Typeahead.js [hunk 11]
```diff
@@ -86,35 +86,40 @@ export class Typeahead extends Component {
...
-          placeholder,
-          onFocus: clearSelection
-        })}
-        suffix={<MatchesLoading isLoading={this.state.calculatingMatches}/>}
+        <div className={listClasses} data-id="report-filters-dropdown">
+          <ActionList actions={mappedMatches} maxHeight={300} />
+        </div>
+        <label className={styles.Label} {...getLabelProps()}>
+          Reports Filter Typeahead
+        </label>
```
Comment (kchu1337) [inline src/pages/reports/components/Typeahead.js]: Added this label

File: src/pages/reports/components/Typeahead.js [hunk 12]
```diff
@@ -86,35 +86,40 @@ export class Typeahead extends Component {
...
     const listClasses = classnames(styles.List, isOpen && mappedMatches.length && styles.open);
     return (
       <div className={styles.Typeahead}>
-        <div className={listClasses}><ActionList actions={mappedMatches} maxHeight={300} /></div>
-        <TextField {...getInputProps({
-          placeholder,
-          onFocus: clearSelection
-        })}
-        suffix={<MatchesLoading isLoading={this.state.calculatingMatches}/>}
+        <div className={listClasses} data-id="report-filters-dropdown">
```
Comment (kchu1337) [inline src/pages/reports/components/Typeahead.js]: Added this data-id

File: src/pages/reports/components/Typeahead.js [hunk 13]
```diff
@@ -86,35 +86,40 @@ export class Typeahead extends Component {
   onTypeahead = ({
     getInputProps,
     getItemProps,
+    getLabelProps,
```
Comment (kchu1337) [inline src/pages/reports/components/Typeahead.js]: added this

File: cypress/tests/integration/analytics/SummmaryReport.spec.js (hunk 3 above)
Comment (nicklemmon) [reply cypress/tests/integration/analytics/SummmaryReport.spec.js]: I was totally wrong - looking at it again, you're good.

File: cypress/tests/integration/analytics/SummmaryReport.spec.js (hunk 6 above)
Comment (nicklemmon) [reply cypress/tests/integration/analytics/SummmaryReport.spec.js]: Nice! Looks great. Nailed it.
